### DISCUSSION TITLE
Aftertouch

### DISCUSSION THREAD

**Discussion by miotislucifugis** [probonopd/MiniDexed#162] [Ideas]
Has aftertouch been implemented yet?   Also, is there a master list of the CC# commands?
Thanks

**Comment by probonopd** [comment]
__Update 11/2023:__

* Monophonic aftertouch is working
* Polyphonic aftertouch is not implemented yet

In Synth Geekery Show episode 223, the subject of polyphonic aftertouch [came up](https://www.youtube.com/live/TDSy5nnm0jA?feature=shared&t=9039).

Currently our source code [says](https://github.com/probonopd/MiniDexed/blob/d9dc84d583e92156a86bbd97f54911f26a2a8014/src/mididevice.cpp#L37)

```
#define MIDI_CHANNEL_AFTERTOUCH 0b1101   // right now Synth_Dexed just manage Channel Aftertouch not Polyphonic AT -> 0b1010
```

What would we need to do to get polyphonic aftertouch going? @dcoredump would https://codeberg.org/dcoredump/Synth_Dexed need to be extended?

**Reply by eLuSiVeMiTe** [reply]
This explains it better than I could. Copied from the moog forum. 
That's because Aftertouch isn't a MIDI Continuous Controller (MIDI CC). Although Aftertouch (AT for short) is a type of controller, it is classified as a separate event type in its own right (like Pitch Bend).

There are two types of AT (Monophonic, which is also called Channel Aftertouch, and Polyphonic). The Voyager produces Monophonic AT.

Monophonic AT is a single data byte message that produces 128 values (low resolution). If you route your Voyager into a MIDI Sequencer program, turn recording on and then dig into the Voyager keyboard, you should be able to record and display the Channel AT data for yourself (provided the software isn't filtering this out).

Of course, what the Voyager (or LP, tone module, etc.) will make of the AT data depends on the preset. On the Voyager, AT is called 'Pressure' and it appears on the two Mod Busses as Shapers. Beyond that, you have to dig into the Pot Mapping menu to program it. If you have a VX-351, you can also route the Pressure CV signal (KDB PRESS on the '351) to any of the Voyager CV inputs.

On the Voyager Old School, Pressure is available as a Controller on Mod Bus 1 & 2.

**Reply by craigyjp** [reply]
The DX7/5 only supported channel aftertouch so it is not too bad. I was testing with my Arturia keystep and noticed aftertouch was not working.  

I think the DX1 had a poly AT keyboard so it must have supported it. But I could be wrong

**Reply by ghost** [reply]
I am owner of a Yamaha KX-88 (the old tank from 1985 https://yamahablackboxes.com/collection/yamaha-kx88-controller/ ) for 12 years now, but sadly the master midi keyboard collected some dust over the last years. Would love to use it the way it was built back then in combination with TX816 modules where the A, B and A + B switching with different channel sounds really shines. As soon as i will get the serial midi port working i can report some infos regarding channel aftertouch, as the KX88 is able to do so. For now i am still trying to get one of my cheap chinese plastic housing built Yamaha KX (KX49 and KX25) from the 2000er years recognized and working. For some strange reason there is no midi data in debug mode showing up, and no sound because of that....

**Comment by miotislucifugis** [comment]
Yeah, channel aftertouch was on the Dx7 and Dexed... Not sure if it is in SynthDexed.   i was browsing the callbacks and didnt see it.    As mentioned, is not a CC#, but its own message type, with a range of 1-127.    I believe on the DX7 there were two parameters- destination and depth.

**Reply by probonopd** [reply]
https://codeberg.org/dcoredump/Synth_Dexed/ (which we are using) has

```
void setAftertouch(uint8_t value);
void setAftertouchRange(uint8_t range);
void setAftertouchTarget(uint8_t target);
```

So the question is: How to wire it up in software. Maybe @dcoredump can give us a hint?

**Reply by craigyjp** [reply]
In the DX7 it's performance parameters I think, rather than a per patch values, but that wouldn't translate too well to a multitimbral synth as each TG needs its own settings for aftertouch.

**Reply by dcoredump** [reply]
MicroDexed handles Aftertouch with a callback handler from the MIDI library.

You can take a look at MicroDexed.ino and search for `handleAftertouch()`.

I can write a more specific text tomorrow.

**Reply by dcoredump** [reply]
> In the DX7 it's performance parameters I think, rather than a per patch values, but that wouldn't translate too well to a multitimbral synth as each TG needs its own settings for aftertouch.

For MicroDexed I added the performance data to the voice data.

**Reply by dcoredump** [reply]
> https://codeberg.org/dcoredump/Synth_Dexed/ (which we are using) has
> 
> ```
> void setAftertouch(uint8_t value);
> void setAftertouchRange(uint8_t range);
> void setAftertouchTarget(uint8_t target);
> ```
> 
> So the question is: How to wire it up in software. Maybe @dcoredump can give us a hint?

You have to check for midi messages with `0xDn` (where n is the channel). The next byte is the aftertouch level, which has to be set by `setAftertouch()`. With `setAftertouchRange()` you can set how much the impact to the target is. With `setAftertouchTarget()` you can set which targets are involved:  0: pitch, 1:amp, 2: EG bias (or combinations by addition of values). You must also have in mind that the voice parameters have also be configured to hear changes.

Aftertouch parameters are so called _performance_ parameters which a DX7 stores global for all voices. It would be better to store them along with the voice data (this is what MicroDexed does).

A good information source is: [https://homepages.abdn.ac.uk/d.j.benson/pages/dx7/sysex-format.txt](https://homepages.abdn.ac.uk/d.j.benson/pages/dx7/sysex-format.txt)

Hope that helps!

**Comment by dialtr** [comment]
I see that polyphonic aftertouch is not supported, but there is a TODO for this in the source code. I wasn't sure if there was interest in supporting this since it was not supported in the original synthesizer, but I own an Ensoniq SQ-80 (which generates poly-at) and am also a C++ programmer with 20+ years of experience. I'd be interested in taking a shot at this if there is interest, but would defer to the maintainers. Thank you!

**Comment by dcoredump** [comment]
Hi Tom @dialtr
 
>  a C++ programmer with 20+ years of experience. I'd be interested in taking a shot at this if there is interest, but would defer to the maintainers. Thank you!

MiniDexed is based on Dexeds sound engine. These engine(s) have been specially adapted for use in microcontrollers and are located in a repository on codeberg ([Synth_Dexed](https://codeberg.org/dcoredump/Synth_Dexed)).

It would be great if the code could be extended to multi-AT (even if I don't have a keyboard that can do multi-AT). The best variant would be, if you make a fork directly on codeberg and create PRs. Then the whole thing could be integrated in the current workflow relatively easy and it can be used for the other current  implementations ([MicroDexed](https://codeberg.org/dcoredump/MicroDexed) and [MicroDexed-touch](https://codeberg.org/positionhigh/MicroDexed-touch))

Let me know if you need more information or if you are looking for a different solution. You can also contact me directly (dcoredump@googlemail.com).

Regards, Holger

**Comment by probonopd** [comment]
It seems that polyphonic aftertouch is not wired up in MiniDexed yet.

I now have a Launchpad X which can do both types of aftertouch (possibly the least expensive option), so I would be able to test it. We can also use a test MIDI file: 
In the attached test MIDI file [aftertouch-channel-poly-test.mid.zip](https://github.com/probonopd/MiniDexed/files/13242959/aftertouch-channel-poly-test.mid.zip)
* The low C is played with channel (monophonic) aftertouch
* The high C is played with polyphonic aftertouch

@dialtr would you still be available/interested in debugging/implementing this?

(Post updated to reflect that monophonic aftertouch is already working.)

**Comment by miotislucifugis** [comment]
Channel aftertouch on MIniDexed is working here.    I haven't messed with the setting recently, so I forget the details, but it always seemed to work properly,( since it was implemented.)   Im using a keystep 37 and a Qunexus   and minidexed using a pi 3

**Reply by probonopd** [reply]
I can confirm that it is working for me now (using performances I made for approximating the DX5 performances).

**Comment by Banana71** [comment]
Then I'll take the opportunity to put my Hydrasynth into operation and recreate a few sounds for the miniDexed with the appropriate modulations.  (Aftertouch, ModWheel, Foot Controller)

**Comment by miotislucifugis** [comment]
I know for sure Electric piano in Rom#1 works- I think that I just enabled AT in the mod menu.   set to pitch for a nice wobble when I press.   range is prob at max.   Do the other mod sources work for you?    have you tried midi monitor (or other midi viewer) to see that you are actually sending AT messages?

**Comment by Banana71** [comment]
### miniDexed Aftertouch Demo
[Sawpressive.zip](https://github.com/probonopd/MiniDexed/files/13248804/Sawpressive.zip)


poly aftertouch doesn't work for me.
Channel aftertouch works as expected.
The performance can be copied directly into the root directory but name your own default performance beforehand otherwise it will be overwritten. Or rename the file to 000nnn_Sawpress.ini and copy it to your Performance folder.
Sawexpressive is a layer of 8 TGs, aftertouch controls pitch and EG bias
What I played wasn't ideal, it was more for modulation with the foot controller.
I slowly increased the pressure on the keys. Not so easy to keep the pressure even.

https://github.com/probonopd/MiniDexed/assets/104296205/d494875a-10b9-4fc2-a2e3-b78f3cd4ae05

**Reply by dcoredump** [reply]
An original DX7 does not support poly AT, so the used Dexed version (on which the fork to Synth_Dexed was based on) does also not support it. Maybe future versions of Dexed will do? But Channel-AT is supported and should do (if the MIDI code recognizes it and forwards it to the engine - which seems to work).

Thanks for testing!
